### PR TITLE
Provide container-aware factories

### PR DESCRIPTION
**Pull request by codeliner**
This PR is not completed yet. 

Missing parts:
- [x] Finish default component configuration
- [x] Write documentation about how to use the factories

### CODE REVIEW

File: config/prooph_service_bus.config.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            //This section will be used by Prooph\ServiceBus\Factory\QueryBusFactory
+            'query_bus' => [
+                //You can add a list of container service ids
+                //The factory will use these to get the plugins from the container
+                'plugins' => []
+
+            ],
+        ]
+    ]
+];
```
Comment (prolic) [inline config/prooph_service_bus.config.php]: grrr :)

File: src/Factory/AbstractBusFactory.php [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        $routerClass = isset($routerConfig['type'])? (string)$routerConfig['type'] : $this->getDefaultRouterClass();
+
+        $routes = isset($routerConfig['routes'])? $routerConfig['routes'] : [];
+
+        $router = new $routerClass($routes);
+
+        $bus->utilize($router);
+    }
+}
+
```
Comment (prolic) [inline src/Factory/AbstractBusFactory.php]: Please check your phpstorm templates ;)

File: src/Factory/AbstractBusFactory.php [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+     * @return array
+     * @throws RuntimeException
+     */
+    private function getBusConfig(ContainerInterface $container)
+    {
+        if (! $container->has('config')) {
+            return [];
+        }
+
+        $config = (array)$container->get('config');
```
Comment (prolic) [inline src/Factory/AbstractBusFactory.php]: Is this really neccessary? If container is an object, it should implement array access.

File: tests/Factory/BusFactoriesTest.php [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    function provideBuses()
+    {
+        return [
+            [CommandBus::class, 'command_bus', new CommandBusFactory()],
+            [EventBus::class, 'event_bus', new EventBusFactory()],
+            [QueryBus::class, 'query_bus', new QueryBusFactory()],
+        ];
+    }
+}
+
```
Comment (prolic) [inline tests/Factory/BusFactoriesTest.php]: space again?

File: config/prooph_service_bus.config.php (hunk 1 above)
Comment (codeliner) [inline config/prooph_service_bus.config.php]: arrrgh, I checked the template, but this f** space is added by PHPStrom. In my defense I have to say that the PR is not ready yet ;). I'll fix it with my next commit. I think I have to investigate some time to set up php-cs-fixer ...

File: src/Factory/AbstractBusFactory.php (hunk 3 above)
Comment (codeliner) [inline src/Factory/AbstractBusFactory.php]: hhhm, I wanted to make sure that `config` is an array. Is the type cast applied to the container? TBH, I didn't tried it so possibly a good catch

Comment (prolic) [inline src/Factory/AbstractBusFactory.php]: In ZF2 config is an instance of Zend\Config\Config which implements ArrayAccess, see:

https://github.com/zendframework/zend-config/blob/master/src/Config.php

I don't know for other frameworks, but it should be an array or ArrayAccess implementation.

File: config/prooph_service_bus.config.php (hunk 1 above)
Comment (prolic) [inline config/prooph_service_bus.config.php]: There is a present for you: https://github.com/prooph/event-store/pull/62

File: src/Factory/AbstractBusFactory.php (hunk 3 above)
Comment (codeliner) [inline src/Factory/AbstractBusFactory.php]: you are right. I change that

File: config/prooph_service_bus.config.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                //To lazy-load query handlers you can provide a service id instead.
+                //In this case the handler is pulled from the container using the provided handler service id
+                'routes' => [],
+                //Router defaults to Prooph\ServiceBus\Plugin\Router\QueryRouter
+                //Comment out the next line to use the RegexRouter instead
+                //'type' => \Prooph\ServiceBus\Plugin\Router\RegexRouter::class,
+            ],
+        ], //EO service_bus
+    ], //EO prooph
+];
```
Comment (prolic) [inline config/prooph_service_bus.config.php]: Travis wird meckern! ;) hehe

Comment (prolic) [inline config/prooph_service_bus.config.php]: Nope. Please add config dir to .php_cs file

**Pull request merged**